logger/logger.go

```diff
@@ -23,29 +22,24 @@ const (
 	traceLevel
 )
 
-// nameToLevel is used to get corresponding level for logger level strings.
-var nameToLevel = map[Level]int{
-	CriticalLevel: criticalLevel,
-	ErrorLevel:    errorLevel,
-	WarnLevel:     warnLevel,
-	InfoLevel:     infoLevel,
-	DebugLevel:    debugLevel,
-	TraceLevel:    traceLevel,
-	OffLevel:      offLevel,
-}
-
-// isValidLogLevel returns true if the given logger level is valid.
-// The check is done case-insensitive.
-func isValidLogLevel(logLevel Level) bool {
-	logLevelStr := strings.ToLower(string(logLevel))
-	_, found := nameToLevel[Level(logLevelStr)]
-	return found
-}
-
 // GetLogLevel returns the corresponding logger level with the given string if it exists, otherwise returns an error.
 func GetLogLevel(logLevel Level) (int, error) {
-	if !isValidLogLevel(logLevel) {
+	switch logLevel {
+	case OffLevel:
+		return offLevel, nil
+	case CriticalLevel:
+		return criticalLevel, nil
+	case ErrorLevel:
+		return errorLevel, nil
+	case WarnLevel:
+		return warnLevel, nil
+	case InfoLevel:
+		return infoLevel, nil
+	case DebugLevel:
+		return debugLevel, nil
+	case TraceLevel:
+		return traceLevel, nil
```

Comment:
Does it make sense to invert the order the levels to pay the least cost the common scenario, such as: Trace, Debug, Info, Warn, Error, Critical, Off? (so getting the Trace level is fastest, while getting the Off level is the slowest, which is not used much).

Reply:
https://play.golang.org/p/LcdE381wsK9
I could not find a meaningful difference. Do you think we can prove this? I would love to learn more about this.
usage: go test --benchmark=. benchmark_test.go

source: https://adayinthelifeof.nl/2021/03/04/go-map-vs-switch.html

If we do not see a performance improvement with the suggested ordering approach, I think it would be more readable as is right now. Wdyt?

Reply:
In the link you've sent it mentions that the switch cases are hold in a linked list, so the cases at the head of the list should be traversed first, hence should be faster to execute (although just a tiny bit). Not sure about the readability. Does ordering the cases in reverse make th switch less readable?

Reply:
Ah okay, I did not read that part.
`go test --bench=. bench_test.go 
goos: darwin
goarch: arm64
BenchmarkForFrequentLogLevels-8                         309533968                3.617 ns/op
BenchmarkForFrequentLogLevelsWithOrderedSwitch-8        330991947                3.628 ns/op
`
Seeing the benchmark results, I assumed it was a jump table. Now I re-run benchmark few times, and it seems this result was a fluke. You are right, I do observe ordered one to be the faster variant. Will fix that, thanks!